Zend/zend_enum.c

```diff
@@ -0,0 +1,350 @@
+	enum_handlers.write_property = zend_enum_write_property;
+	enum_handlers.unset_property = zend_enum_unset_property;
+	enum_handlers.get_property_ptr_ptr = zend_enum_get_property_ptr_ptr;
+	enum_handlers.clone_obj = NULL;
+	enum_handlers.compare = zend_enum_compare_objects;
+}
+
+void zend_enum_add_interfaces(zend_class_entry *ce)
+{
+	uint32_t num_interfaces_before = ce->num_interfaces;
+
+	ce->num_interfaces++;
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		ce->num_interfaces++;
+	}
+
+	ce->interface_names = erealloc(ce->interface_names, sizeof(zend_class_name) * ce->num_interfaces);
+
+	ce->interface_names[num_interfaces_before].name = zend_string_init("UnitEnum", sizeof("UnitEnum") - 1, 0);
+	ce->interface_names[num_interfaces_before].lc_name = zend_string_init("unitenum", sizeof("unitenum") - 1, 0);
+
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		ce->interface_names[num_interfaces_before + 1].name = zend_string_init("ScalarEnum", sizeof("ScalarEnum") - 1, 0);
+		ce->interface_names[num_interfaces_before + 1].lc_name = zend_string_init("scalarenum", sizeof("scalarenum") - 1, 0);	
+	}
+}
+
+static ZEND_NAMED_FUNCTION(zend_enum_cases_func)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_class_constant *c;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	array_init(return_value);
+
+	ZEND_HASH_FOREACH_PTR(&ce->constants_table, c) {
+		if (!(c->const_flags & ZEND_CLASS_CONST_IS_CASE)) {
+			continue;
+		}
+		zval *zv = &c->value;
+		if (Z_TYPE_P(zv) == IS_CONSTANT_AST) {
+			zval_update_constant_ex(zv, c->ce);
+			if (UNEXPECTED(EG(exception) != NULL)) {
+				return;
+			}
+		}
+		if (ce->enum_scalar_type == IS_UNDEF) {
+			zend_hash_next_index_insert(Z_ARRVAL_P(return_value), zv);
+		} else {
+			zend_object *zobj = Z_OBJ_P(zv);
+			zend_string *value_string = zend_string_init("value", strlen("value"), 0);
+			zval tmp;
+			zval *value_zv = zend_read_property_ex(ce, zobj, value_string, 0, &tmp);
+			zend_string_free(value_string);
+
+			if (ce->enum_scalar_type == IS_LONG) {
+				zend_hash_index_update(Z_ARRVAL_P(return_value), Z_LVAL_P(value_zv), zv);
+			} else {
+				ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+				zend_symtable_update(Z_ARRVAL_P(return_value), Z_STR_P(value_zv), zv);
+			}
+		}
+	} ZEND_HASH_FOREACH_END();
+}
+
+static ZEND_NAMED_FUNCTION(zend_enum_from_func)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_string *string_key;
+	zend_long long_key;
+
+	zval *case_name_zv;
+	if (ce->enum_scalar_type == IS_LONG) {
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &long_key) == FAILURE) {
+			RETURN_THROWS();
+		}
+
+		case_name_zv = zend_hash_index_find(ce->enum_scalar_table, long_key);
+	} else {
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "S", &string_key) == FAILURE) {
+			RETURN_THROWS();
+		}
+
+		ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+		case_name_zv = zend_hash_find_ex(ce->enum_scalar_table, string_key, 1);
```

Comment:
```suggestion
		case_name_zv = zend_hash_find(ce->enum_scalar_table, string_key);
```

Similarly, the hash isn't known here, so passing known_hash=1 may incorrectly throw in from() for dynamic string values

I think compilation uses interned strings, so the code in zend_compile.c is fine?